openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,284 @@
+"""Define the SubproblemComp class for evaluating OpenMDAO systems within problems."""
+
+import openmdao.api as om
+from openmdao.core.constants import _UNDEFINED
+from openmdao.utils.om_warnings import issue_warning
+from openmdao.core.driver import Driver
+
+
+def _get_model_vars(varType, vars, model_vars):
+    """
+    Get the requested IO variable data from model's list of IO.
+
+    Parameters
+    ----------
+    varType : str
+        Specifies whether inputs or outputs are being extracted.
+    vars : list of str or tuple
+        List of provided variable names in str or tuple form. If an element is a str,
+        then it should be the var name in its promoted name. If it is a tuple, then the
+        first element should be the promoted name, and the second element should be the
+        var name you wish to refer to it by within the subproblem [e.g. (prom_name, var_name)].
+    model_vars : dict
+        Dict of model's IO vars and meta data.
+
+    Returns
+    -------
+    dict
+        Dict to update `self.options` with desired IO data in `SubproblemComp`.
+    """
+    var_dict = {varType: {}}
+
+    for var in vars:
+        if isinstance(var, tuple):
+            # check if variable already exists in var_dict[varType]
+            # i.e. no repeated variable names
+            if var[1] in var_dict[varType]:
+                raise Exception(f'Variable {var[1]} already exists. Rename variable'
+                                ' or delete copy of variable.')
+
+            # make dict with given var name as key and meta data from model_vars
+            tmp_dict = {var[1]: meta for _, meta in model_vars.items()
+                        if meta['prom_name'] == var[0]}
+
+            # check if dict is empty (no vars added)
+            if len(tmp_dict) == 0:
+                raise Exception(f'Promoted name {var[0]} does not'
+                                ' exist in model.')
+
+            var_dict[varType].update(tmp_dict)
+
+        elif isinstance(var, str):
+            # check if variable already exists in var_dict[varType]
+            # i.e. no repeated variable names
+            if var in var_dict[varType]:
+                raise Exception(f'Variable {var} already exists. Rename variable'
+                                ' or delete copy of variable.')
+
+            # make dict with given var name as key and meta data from model_vars
+            tmp_dict = {var: meta for _, meta in model_vars.items()
```

Comment:
Not sure why this one was resolved without comment or code change...

Reply:
I changed it from 
```python
tmp_dict = {var: meta for _, meta in model_vars
                        if name.endswith(var)}
```
to
```python
tmp_dict = {var: meta for name, meta in model_vars
                        if name.endswith('.' + var)}
```
so that it catches what you mentioned with `prom_name`. Along with that, I made sure `get_model_io` was passed the absolute names rather than the promoted names so that there will always be a '.' in the name since sometimes `prom_name` won't include the '.' syntax.

Reply:
If you'd rather work with promoted names, you could change the check to `if name==var or name.endswith('.' + var)`.  You do have to be a little careful here, making sure that you're comparing the same type of names, i.e. you can compare promoted to promoted, but not absolute to promoted because there's not necessarily any correspondence between a promoted name and the suffix of an absolute name since promoted names are allowed to be aliased.

Reply:
I may have missed it, but I didn't see any promotion aliases in your test case.  You should probably have a test that includes that.

Reply:
For that check, would adding `meta['prom_name'] == var` to the check you gave before suffice since that adds the option that the `prom_name` is checked along with the absolute name?

Reply:
fixed name check and ensured model.list_inputs/outputs uses absolute names for keys.